### ISSUE TITLE
Unexpected change in behavior causing TeamCity issue

### ISSUE THREAD

**Issue by santhonisz** [GitTools/GitVersion#1874]
Hi there,

We have started encountering failures in our TeamCity builds due to an apparent change in behavior of GitVersion since updating to the 5.1.1 release. It seems that the normalization process is now performing a git fetch by default, whereas in previous versions it did not. This is then causing an issue as we use SSH on our GitHub repositories. I am aware of the `nofetch` and new `nonormalize` switches, but I'm wondering if this change in behavior is intentional or is a regression issue?

**GitVersion 5.0.1**
```[16:40:28]	Running: C:\ProgramData\chocolatey\bin\gitversion.exe "C:\buildAgent\work\44204f131946980e" /p '*****' /output 'buildserver'
[16:40:29]	INFO [10/28/19 16:40:29:01] Working directory: C:\buildAgent\work\44204f131946980e
[16:40:29]	INFO [10/28/19 16:40:29:02] IsDynamicGitRepository: False
[16:40:29]	INFO [10/28/19 16:40:29:04] Returning Project Root from DotGitDirectory: C:\buildAgent\work\44204f131946980e\.git - C:\buildAgent\work\44204f131946980e
[16:40:29]	INFO [10/28/19 16:40:29:04] Running on Windows.
[16:40:29]	INFO [10/28/19 16:40:29:04] Applicable build agent found: 'TeamCity'.
[16:40:29]	INFO [10/28/19 16:40:29:04] Branch from build environment: refs/pull/103/head
[16:40:29]	INFO [10/28/19 16:40:29:04] Begin: Normalizing git directory for branch 'refs/pull/103/head'
[16:40:29]	  INFO [10/28/19 16:40:29:08] One remote found (origin -> 'git@github.com:oneiress/iosplus-client.git').
[16:40:29]	  INFO [10/28/19 16:40:29:08] Skipping fetching, if GitVersion does not calculate your version as expected you might need to allow fetching or use dynamic repositories
```

**GitVersion 5.1.1**
```[10:28:58]	Running: C:\ProgramData\chocolatey\bin\gitversion.exe "C:\buildAgent\work\44204f131946980e" /p '*****' /output 'buildserver'
[10:28:59]	INFO [10/29/19 10:28:59:30] Working directory: C:\buildAgent\work\44204f131946980e
[10:28:59]	INFO [10/29/19 10:28:59:31] IsDynamicGitRepository: False
[10:28:59]	INFO [10/29/19 10:28:59:33] Returning Project Root from DotGitDirectory: C:\buildAgent\work\44204f131946980e\.git\ - C:\buildAgent\work\44204f131946980e\
[10:28:59]	INFO [10/29/19 10:28:59:33] Running on Windows.
[10:28:59]	INFO [10/29/19 10:28:59:33] Applicable build agent found: 'TeamCity'.
[10:28:59]	INFO [10/29/19 10:28:59:33] Branch from build environment: refs/pull/104/head
[10:28:59]	INFO [10/29/19 10:28:59:33] Begin: Normalizing git directory for branch 'refs/pull/104/head'
[10:28:59]	  INFO [10/29/19 10:28:59:36] One remote found (origin -> 'git@github.com:oneiress/iosplus-client.git').
[10:28:59]	  INFO [10/29/19 10:28:59:36] Fetching from remote 'origin' using the following refspecs: +refs/heads/*:refs/remotes/origin/*.
[10:28:59]	  INFO [10/29/19 10:28:59:37] End: Normalizing git directory for branch 'refs/pull/104/head' (Took: 36.66ms)
[10:28:59]	  INFO [10/29/19 10:28:59:39] 
[10:28:59]	  ERROR [10/29/19 10:28:59:39] An unexpected error occurred:
[10:28:59]	  INFO [10/29/19 10:28:59:39] Attempting to show the current git graph (please include in issue): 
[10:28:59]	LibGit2Sharp.LibGit2SharpException: unsupported URL protocol
[10:28:59]	  INFO [10/29/19 10:28:59:39] Showing max of 100 commits
[10:28:59]	   at LibGit2Sharp.Core.Ensure.HandleError(Int32 result)
[10:28:59]	   at LibGit2Sharp.Core.Ensure.ZeroResult(Int32 result)
[10:28:59]	   at LibGit2Sharp.Core.Proxy.git_remote_fetch(RemoteHandle remote, IEnumerable`1 refSpecs, GitFetchOptions fetchOptions, String logMessage)
[10:28:59]	   at LibGit2Sharp.Commands.Fetch(Repository repository, String remote, IEnumerable`1 refspecs, FetchOptions options, String logMessage)
[10:28:59]	   at GitVersion.Helpers.GitRepositoryHelper.Fetch(ILog log, AuthenticationInfo authentication, Remote remote, Repository repo)
[10:28:59]	   at GitVersion.Helpers.GitRepositoryHelper.NormalizeGitDirectory(ILog log, IEnvironment environment, String gitDirectory, AuthenticationInfo authentication, Boolean noFetch, String currentBranch, Boolean isDynamicRepository)
[10:28:59]	   at GitVersion.GitPreparer.Prepare(Boolean normalizeGitDirectory, String currentBranch, Boolean shouldCleanUpRemotes)
[10:28:59]	   at GitVersion.GitVersionCalculator.CalculateVersionVariables()
[10:28:59]	   at GitVersion.ExecCommand.Execute()
[10:28:59]	   at GitVersion.GitVersionExecutor.VerifyArgumentsAndRun(Arguments arguments)
```

Thanks,

Sam

**Comment by mderriey** [comment]
Hi there 👋

We're experiencing the same issue from an Azure Pipelines build.  
Let us know if we can provide any more information to help.

```
INFO [10/29/19 2:49:37:45] Applicable build agent found: 'AzurePipelines'.
INFO [10/29/19 2:49:37:46] Working directory: C:\agent\_work\7\sINFO [10/29/19 2:49:37:46] IsDynamicGitRepository: False
INFO [10/29/19 2:49:37:50] Returning Project Root from DotGitDirectory: C:\agent\_work\7\s\.git\ - C:\agent\_work\7\s\
INFO [10/29/19 2:49:37:50] Running on Windows.INFO [10/29/19 2:49:37:51] Applicable build agent found: 'AzurePipelines'.
INFO [10/29/19 2:49:37:51] Branch from build environment: refs/heads/set-office-property-in-azuread
INFO [10/29/19 2:49:37:51] Begin: Normalizing git directory for branch 'refs/heads/set-office-property-in-azuread'
INFO [10/29/19 2:49:37:60] One remote found (origin -> '<url-of-origin-remote>').
INFO [10/29/19 2:49:37:61] Fetching from remote 'origin' using the following refspecs: +refs/heads/*:refs/remotes/origin/*.
INFO [10/29/19 2:49:37:93] End: Normalizing git directory for branch 'refs/heads/set-office-property-in-azuread' (Took: 424.13ms)
ERROR [10/29/19 2:49:37:96] An unexpected error occurred:
LibGit2Sharp.LibGit2SharpException: request failed with status code: 401
   at LibGit2Sharp.Core.Ensure.HandleError(Int32 result)
   at LibGit2Sharp.Core.Ensure.ZeroResult(Int32 result)
   at LibGit2Sharp.Core.Proxy.git_remote_fetch(RemoteHandle remote, IEnumerable`1 refSpecs, GitFetchOptions fetchOptions, String logMessage)
   at LibGit2Sharp.Commands.Fetch(Repository repository, String remote, IEnumerable`1 refspecs, FetchOptions options, String logMessage)
   at GitVersion.Helpers.GitRepositoryHelper.Fetch(ILog log, AuthenticationInfo authentication, Remote remote, Repository repo)
   at GitVersion.Helpers.GitRepositoryHelper.NormalizeGitDirectory(ILog log, IEnvironment environment, String gitDirectory, AuthenticationInfo authentication, Boolean noFetch, String currentBranch, Boolean isDynamicRepository)
   at GitVersion.GitPreparer.Prepare(Boolean normalizeGitDirectory, String currentBranch, Boolean shouldCleanUpRemotes)
   at GitVersion.GitVersionCalculator.CalculateVersionVariables()
   at GitVersion.ExecCommand.Execute()
   at GitVersion.GitVersionExecutor.VerifyArgumentsAndRun(Arguments arguments)
```

**Comment by davidkeaveny** [comment]
I am experiencing this issue with GitVersion.Portable 5.1.1, running on TeamCity with the repository in BitBucket Cloud:, using SSH as the protocol.

```
INFO [10/29/19 16:02:23:93] Begin: Normalizing git directory for branch 'refs/heads/master'
INFO [10/29/19 16:02:24:19] One remote found (origin -> 'ssh://myproject@bitbucket.org/myproject/api.git').
INFO [10/29/19 16:02:24:19] Fetching from remote 'origin' using the following refspecs: +refs/heads/*:refs/remotes/origin/*.
INFO [10/29/19 16:02:24:21] End: Normalizing git directory for branch 'refs/heads/master' (Took: 280.89ms)
ERROR [10/29/19 16:02:24:24] An unexpected error occurred:
INFO [10/29/19 16:02:24:24] 
INFO [10/29/19 16:02:24:24] Attempting to show the current git graph (please include in issue): 
  LibGit2Sharp.LibGit2SharpException: unsupported URL protocol
INFO [10/29/19 16:02:24:24] Showing max of 100 commits
  at LibGit2Sharp.Core.Ensure.HandleError(Int32 result)
  at LibGit2Sharp.Core.Ensure.ZeroResult(Int32 result)
  at LibGit2Sharp.Core.Proxy.git_remote_fetch(RemoteHandle remote, IEnumerable`1 refSpecs, GitFetchOptions fetchOptions, String logMessage)
  at LibGit2Sharp.Commands.Fetch(Repository repository, String remote, IEnumerable`1 refspecs, FetchOptions options, String logMessage)
  at GitVersion.Helpers.GitRepositoryHelper.Fetch(ILog log, AuthenticationInfo authentication, Remote remote, Repository repo)
  at GitVersion.Helpers.GitRepositoryHelper.NormalizeGitDirectory(ILog log, IEnvironment environment, String gitDirectory, AuthenticationInfo authentication, Boolean noFetch, String currentBranch, Boolean isDynamicRepository)
  at GitVersion.GitPreparer.Prepare(Boolean normalizeGitDirectory, String currentBranch, Boolean shouldCleanUpRemotes)
  at GitVersion.GitVersionCalculator.CalculateVersionVariables()
  at GitVersion.ExecCommand.Execute()
  at GitVersion.GitVersionExecutor.VerifyArgumentsAndRun(Arguments arguments)
INFO [10/29/19 16:02:24:52] * 0bd20129 10 minutes ago  (HEAD -> master, origin/master)
```

**Comment by davidkeaveny** [comment]
In my case, downgrading to 5.0.1 and clearing the auto-update setting in TeamCity avoids the issue.

**Comment by jakubsuchybio** [comment]
#MeToo

**Comment by Bishbulb** [comment]
#MeToo. All of our TC builds suddenly started failing yesterday. Going to try @davidkeaveny's suggestion and downgrade and disable the auto-update setting.

**Comment by bencameron** [comment]
I'm getting the same issue in Azure DevOps.  We're using GitVersion.CommandLine within a Cake script and started seeing the same errors yesterday.  We've updated a few critical build scripts to pin to the 5.0.1 version, but we're waiting to hear if this new behavior is permanent before applying a fix to all of our other scripts.

Does anyone know if this behavior is here to stay?

**Comment by claudiuconstantin** [comment]
I also encountered this, the temporary solution was to downgrade to v5.0.1

**Comment by mderriey** [comment]
Hey @arturcic, I'm really not familiar with the code base, but if you have pointers, I should have some time next week to try and have a look at this.

**Comment by arturcic** [comment]
@santhonisz , @mderriey can you check version 5.1.2-beta.1+4 if it fixed the error?

**Comment by mderriey** [comment]
Hi @arturcic, I tried with both `5.1.2-beta.1+4` and `5.1.2-beta.1+5` and while the error about the fetch went away, it fails to compute a version when running from a directory where Azure DevOps cloned the repository for a pull-request CI build, so it checks out a specific commit by SHA, where this commits represents the merge of the feature branch with `master`.

See the logs below:
```
INFO [11/03/19 23:23:26:02] Working directory: C:\agent\_work\7\s
INFO [11/03/19 23:23:26:03] IsDynamicGitRepository: False
INFO [11/03/19 23:23:26:07] Returning Project Root from DotGitDirectory: C:\agent\_work\7\s\.git\ - C:\agent\_work\7\s\
INFO [11/03/19 23:23:26:08] Running on Windows.
INFO [11/03/19 23:23:26:08] IsDynamicGitRepository: False
INFO [11/03/19 23:23:26:08] Returning Project Root from DotGitDirectory: C:\agent\_work\7\s\.git\ - C:\agent\_work\7\s\
INFO [11/03/19 23:23:26:08] Project root is: C:\agent\_work\7\s\
INFO [11/03/19 23:23:26:08] DotGit directory is: C:\agent\_work\7\s\.git
INFO [11/03/19 23:23:26:16] IsDynamicGitRepository: False
INFO [11/03/19 23:23:26:16] Returning Project Root from DotGitDirectory: C:\agent\_work\7\s\.git\ - C:\agent\_work\7\s\
INFO [11/03/19 23:23:26:21] Begin: Loading version variables from disk cache  
INFO [11/03/19 23:23:26:21] Cache file C:\agent\_work\7\s\.git\gitversion_cache\C8BBB8D40632C11BE1C6B133DE3AEAFF9106AED3.yml not found.  
INFO [11/03/19 23:23:26:22] End: Loading version variables from disk cache (Took: 4.18ms)  
INFO [11/03/19 23:23:26:22] IsDynamicGitRepository: False  
INFO [11/03/19 23:23:26:22] Returning Project Root from DotGitDirectory: C:\agent\_work\7\s\.git\ - C:\agent\_work\7\s\  
INFO [11/03/19 23:23:26:26] Using latest commit on specified branch  
INFO [11/03/19 23:23:26:28] Begin: Getting branches containing the commit '4767bde65bba68ead1d1f9d183ea4e46e73cd6fd'.    
INFO [11/03/19 23:23:26:28] Trying to find direct branches.    
INFO [11/03/19 23:23:26:28] No direct branches found, searching through tracked branches.    
INFO [11/03/19 23:23:26:28] Searching for commits reachable from '<reacted-branch-name>'.    
INFO [11/03/19 23:23:26:32] The branch '<reacted-branch-name>' has no matching commits.    
INFO [11/03/19 23:23:26:32] Searching for commits reachable from 'master'.    
INFO [11/03/19 23:23:26:34] The branch 'master' has a matching commit.    
INFO [11/03/19 23:23:26:34] Searching for commits reachable from 'origin/<reacted-branch-name>'.    
INFO [11/03/19 23:23:26:36] The branch 'origin/<reacted-branch-name>' has no matching commits.    
INFO [11/03/19 23:23:26:36] Searching for commits reachable from 'origin/master'.    
INFO [11/03/19 23:23:26:37] The branch 'origin/master' has a matching commit.  
INFO [11/03/19 23:23:26:37] End: Getting branches containing the commit '4767bde65bba68ead1d1f9d183ea4e46e73cd6fd'. (Took: 94.03ms)  
INFO [11/03/19 23:23:26:47] No branch configuration found for branch (no branch), falling back to default configuration  
INFO [11/03/19 23:23:26:48] Begin: Attempting to inherit branch configuration from parent branch    
INFO [11/03/19 23:23:26:50] Begin: Finding branch source of '(no branch)'      
INFO [11/03/19 23:23:26:54] Begin: Finding merge base between '(no branch)' and 'master'.        
INFO [11/03/19 23:23:26:56] Found merge base of 4767bde65bba68ead1d1f9d183ea4e46e73cd6fd        
INFO [11/03/19 23:23:26:56] Merge base of (no branch)' and 'master is 4767bde65bba68ead1d1f9d183ea4e46e73cd6fd      
INFO [11/03/19 23:23:26:56] End: Finding merge base between '(no branch)' and 'master'. (Took: 20.69ms)      
INFO [11/03/19 23:23:26:56] Begin: Finding merge base between '(no branch)' and 'origin/master'.        
INFO [11/03/19 23:23:26:56] Found merge base of 4767bde65bba68ead1d1f9d183ea4e46e73cd6fd        
INFO [11/03/19 23:23:26:56] Merge base of (no branch)' and 'origin/master is 4767bde65bba68ead1d1f9d183ea4e46e73cd6fd      
INFO [11/03/19 23:23:26:56] End: Finding merge base between '(no branch)' and 'origin/master'. (Took: 0.20ms)      
INFO [11/03/19 23:23:26:58] Multiple source branches have been found, picking the first one (master).
This may result in incorrect commit counting.
Options were:
 master, origin/master    
 INFO [11/03/19 23:23:26:58] End: Finding branch source of '(no branch)' (Took: 80.06ms)    
 INFO [11/03/19 23:23:26:58] Begin: Getting branches containing the commit '4767bde65bba68ead1d1f9d183ea4e46e73cd6fd'.      
 INFO [11/03/19 23:23:26:58] Trying to find direct branches.      
 INFO [11/03/19 23:23:26:58] No direct branches found, searching through tracked branches.      
 INFO [11/03/19 23:23:26:58] Searching for commits reachable from 'master'.      
 INFO [11/03/19 23:23:26:60] The branch 'master' has a matching commit.      
 INFO [11/03/19 23:23:26:60] Searching for commits reachable from 'origin/master'.      
 INFO [11/03/19 23:23:26:61] The branch 'origin/master' has a matching commit.    
 INFO [11/03/19 23:23:26:61] End: Getting branches containing the commit '4767bde65bba68ead1d1f9d183ea4e46e73cd6fd'. (Took: 30.55ms)    
 INFO [11/03/19 23:23:26:61] Begin: Getting branches containing the commit '4767bde65bba68ead1d1f9d183ea4e46e73cd6fd'.      
 INFO [11/03/19 23:23:26:61] Trying to find direct branches.      
 INFO [11/03/19 23:23:26:61] No direct branches found, searching through tracked branches.      
 INFO [11/03/19 23:23:26:61] Searching for commits reachable from 'master'.      
 INFO [11/03/19 23:23:26:63] The branch 'master' has a matching commit.      
 INFO [11/03/19 23:23:26:63] Searching for commits reachable from 'origin/master'.      
 INFO [11/03/19 23:23:26:64] The branch 'origin/master' has a matching commit.    
 INFO [11/03/19 23:23:26:64] End: Getting branches containing the commit '4767bde65bba68ead1d1f9d183ea4e46e73cd6fd'. (Took: 29.38ms)    
 INFO [11/03/19 23:23:26:65] Found possible parent branches:   
 WARN [11/03/19 23:23:26:65] Failed to inherit Increment branch configuration, no branches found.

Falling back to master branch config  
INFO [11/03/19 23:23:26:65] End: Attempting to inherit branch configuration from parent branch (Took: 171.71ms)  
INFO [11/03/19 23:23:26:67] Running against branch: (no branch) (4767bde65bba68ead1d1f9d183ea4e46e73cd6fd)
WARN [11/03/19 23:23:26:68] An error occurred:
It looks like the branch being examined is a detached Head pointing to commit '4767bde'. Without a proper branch name GitVersion cannot determine the build version.
```

**Comment by mderriey** [comment]
We installed `5.1.2-beta.1+5` and it worked fine without the `/nonormalize` flag. Thanks heaps!

**Comment by arturcic** [comment]
Closing as there's a confirmation it worked

**Comment by rmajid2** [comment]
Usage of GitVersion as script in Azure Pipelines **Hosted Windows 2019 with VS2019 (v20191028.1)** throw exception ([As v20191028.1 include V5.1.1.0 of GitVersion](https://github.com/microsoft/azure-pipelines-image-generation/blob/vs2019/20191028.1/images/win/Vs2019-Server2019-Readme.md#gitversion)):
`LibGit2Sharp.LibGit2SharpException: request failed with status code: 401`

A workaround and a temporary solution was to downgrade to v5.0.1 by using the [GitVersion task](https://marketplace.visualstudio.com/items?itemName=gittools.gitversion) wich use v5.0.1 of GitVersion

**Comment by arturcic** [comment]
@rmajid2 5.1.1 has several bugs. We fixed most of them in 5.1.2 so you could use 5.1.2 witht eh GitVersion Task extension. Hopefully the Azure Pipeline image will get updated soon
